### PR TITLE
Prevent Wait Animation from adding mobject to scene

### LINKED ISSUE #3040: Scene.wait (and .pause) calls add an empty Mobject to self.mobjects
## Description of bug / unexpected behavior
After each call to Scene's pause() method, I notice a mobject being added to the scene.

```py
class PauseMobjects(Scene):
    def construct(self):
        print(self.mobjects)
        yellow = Square()

        self.play(FadeIn(yellow))
        print(self.mobjects)

        for i in range(0, 5):
            self.pause()
            print(self.mobjects)
```
Console while running above:
```
[]
[Square]
[Square, Mobject]
[Square, Mobject, Mobject]
[Square, Mobject, Mobject, Mobject]
[Square, Mobject, Mobject, Mobject, Mobject]
[Square, Mobject, Mobject, Mobject, Mobject, Mobject]
[Square, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject]
[Square, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject]
[Square, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject]
[Square, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject]
[Square, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject, Mobject]
```

Nothing is rendered here, but I worry that these excess mobjects could be reducing performance, especially because hasing the scene is O(n) where n is the number of mobjects in the scene (e.g. #2710 )

## Expected behavior
I would not expect pausing or otherwise waiting to change the mobjects in the scene. From what I've poked around, it's not apparent to me that adding a Mobject is important for waits, so hopefully this can be removed.

## Discussion
I am working on a fix, trying to figure out where this mobject is added and prevent that, or at least make sure the mobject is cleaned up after the animation finishes.

This is with v0.16.0

### PR DESCRIPTION
**Pull request by kjlubick**
Fixes #3040

I added in some code to give the name of the Animation class to the default Mobject that is there. Hopefully this makes future mystery Mobject less mysterious to track down.

 - `Wait` animations (e.g. `scene.pause()`) no longer add an empty Mobject to the scene.
 - `Group`s throw a `ValueError` if they contain `None`. This would have caused issues before, but the error is now more explicit.

### CODE REVIEW

**Comment by huguesdevimeux** [comment]
Hello and thank you for the Pull Request ! 

Your fix looks like a good idea to me. 
An issue I see is that `scene.add(mobject)` raises an error when `mobject` is `None`. 
This causes issue when there is no check for **None-ity of the mobject**, for example, [here](https://github.com/ManimCommunity/manim/blob/42d595ac4fb6c607a8375e28f24b9b7193668cb5/manim/mobject/mobject.py#L2804-L2805)
https://github.com/ManimCommunity/manim/blob/42d595ac4fb6c607a8375e28f24b9b7193668cb5/manim/mobject/mobject.py#L2792-L2804

But, it works well in the general case because `scene.add_mobject_from_animation` (called for every animation) does this check (l 486)
https://github.com/ManimCommunity/manim/blob/42d595ac4fb6c607a8375e28f24b9b7193668cb5/manim/scene/scene.py#L478-L488

So, I think implementing this "check" everywhere it's needed should resolve the problem, and I hope also the tests.

Thanks a lot for that !

File: manim/animation/animation.py [hunk 1]
```diff
@@ -156,8 +156,11 @@ def __init__(
                 mobject if mobject is not None else OpenGLMobject()
             )
         else:
-            self.starting_mobject: Mobject = Mobject()
-            self.mobject: Mobject = mobject if mobject is not None else Mobject()
+            self.starting_mobject: Mobject = Mobject(name=self.__class__.__name__)
+            if mobject:
+                self.mobject: Mobject = mobject
+            else:
+                self.mobject: Mobject = self.starting_mobject
```
Comment (huguesdevimeux) [inline manim/animation/animation.py:163]: **Nitpick** : This would be better in one line :)

**Review by huguesdevimeux — CHANGES_REQUESTED** [review]

**Comment by kjlubick** [comment]
I'm not sure what you mean about the tests failing due to adding a `None` `Mobject`. As far as I can tell, the logs indicate a missing opengl Python package. Am I misunderstanding those failures?

The only place I could find the animation.mobject being used was in [`Scene.add_mobjects_from_animations`](https://github.com/ManimCommunity/manim/blob/ebe36178e722c8590a44e23872208ee52bc12bbd/manim/scene/scene.py#L485) called by [`Scene.compile_animation_data`](https://github.com/ManimCommunity/manim/blob/ebe36178e722c8590a44e23872208ee52bc12bbd/manim/scene/scene.py#L1140). You'll see there already exists a check for a `None` `Animation.mobject`, which is why I decided to set it to `None `where I did.

Are there other places the Animation's mobject is used outside of the Animation class?

**Comment by kjlubick** [comment]
Ah, I see those failures now. I'll try to resolve that tomorrow

**Comment by huguesdevimeux** [comment]
> Are there other places the Animation's mobject is used outside of the Animation class?

As in my previous message, yes, there is at least one in the `Group`. I just realized that I mixed up `scene.add(mobject)` and `mobject.add(mobject)`. The first one does not raise an error when None is passed as a parameter (although an error will eventually be raised later in the process ...), while the latter does.

**Comment by huguesdevimeux** [comment]
Also, when you are done with your fix, please add a small test for that, thanks :D

**Comment by kjlubick** [comment]
New test added and existing tests (hopefully) fixed. I'm having a hard time running all the tests locally, so it's a bit awkward for me to verify.

File: manim/animation/animation.py (hunk 1 above)
Comment (kjlubick) [reply manim/animation/animation.py]: Done.

**Comment by kjlubick** [comment]
Looks like some (all?) of the remaining failures are due to very small diffs in image tests using wait.

image
image

I got my dev environment working, so I can re-generate the expectations for these 30-40 tests

**Comment by kjlubick** [comment]
Note, we might have to re-update the expectations after #3042 lands because that test includes a Scene.pause()

File: manim/mobject/mobject.py [hunk 2]
```diff
@@ -2801,6 +2801,7 @@ class Group(Mobject, metaclass=ConvertToOpenGL):
 
     def __init__(self, *mobjects, **kwargs):
         super().__init__(**kwargs)
+        assert None not in mobjects
```
Comment (huguesdevimeux) [inline manim/mobject/mobject.py:2804]: Please add an exception here instead of an assertion. a ValueError with an explicit message would be better IMO.

**Comment by huguesdevimeux** [comment]
The failing tests are really, really weird. But I guess that's part of the beauty of Manim to have such obscure side effects. I'm fine with just regenerating the test data, as it does not look to be such a big deal issue.

i added a last suggestion, otherwise everything SGTM :D

**Review by huguesdevimeux — APPROVED** [review]

**Review by behackl — COMMENTED** [review]
Thanks! This looks good to me in general, I'd like to spend a bit of time investigating where the graphical differences come from. And just to avoid breaking things in a weird way, I propose merging this early in the next release cycle.

**Comment by kjlubick** [comment]
Here's a repro of the "drawing things twice while paused" issue:
```python
class PauseBug(Scene):
    def construct(self):
        self.camera.background_color = WHITE

        # These boxes should be the same color, but they are not
        grey1 = Square(fill_color="#888888", fill_opacity=1.0)
        grey2 = Square(fill_color="#000000", fill_opacity=0.5)

        grey2.next_to(grey1, RIGHT)

        self.add(grey1, grey2)
        self.pause()

        text = Text("Animating", font_size=18, color=BLACK).scale(2)
        text.move_to([0, 3, 0])
        self.play(FadeIn(text), run_time=2.0)

        self.remove(text)
        self.wait(2.0)
        self.add(text)

        self.play(FadeOut(text), run_time=2.0)

        self.pause()
```

https://user-images.githubusercontent.com/6819944/208914989-b5418ed1-1f92-47bd-b23f-5cf94a472328.mp4

**Review by behackl — CHANGES_REQUESTED** [review]
Marking this as a draft and requesting changes until the problem with this is fixed (with reasonably high probability I'll have time to investigate further later this week)l.

**Comment by kjlubick** [comment]
Investigation notes

Without the patch, `scene.moving_mobjects` in [cairo_renderer.py](https://github.com/ManimCommunity/manim/blob/3713ed6b1e0f9d51de8e74717bf4489c592f880f/manim/renderer/cairo_renderer.py#L99) is not empty (it has the Wait Mobject).

With the patch, `scene.moving_mobjects` **is** empty, causing [`update_frame`](https://github.com/ManimCommunity/manim/blob/3713ed6b1e0f9d51de8e74717bf4489c592f880f/manim/renderer/cairo_renderer.py#L135-L139) to fill the mobjects parameter with all mobjects in the scene, causing them (?) to be double drawn.

It looks like the condition in update_frame changed from checking for `None` to be `None` or empty in this PR: https://github.com/ManimCommunity/manim/pull/1736

We might want to:

1. revisit the solution to that PR
2. see why Wait is being dropped from the "moving" mobjects
3. have cairo_renderer.play not call scene.update_frame if moving mobjects is empty

Solution 3 seems like the easiest fix, but I'm out of time this morning to investigate further (and see if it actually works).

**Comment by kjlubick** [comment]
I went with option 3 and while the test case from https://github.com/ManimCommunity/manim/pull/3041#issuecomment-1361306591 looks as it should, I've doubled the amount of expectations that changed.

**Comment by JasonGrace2282** [comment]
This PR is almost 2 years old now, and the initial issue of `Wait` adding a mobject should be fixed (soon) in the experimental branch.
As such, I'll close the PR.

**Pull request closed without merging**